spec_02.md

```diff
@@ -0,0 +1,86 @@
+- Spec: 2
+- Title: Methods on figure objects
+- Version: 0.1
+- Last-Modified: 2019-07-19
+- Authors: [James Bednar](),
+           [David Hoese](),
+           [Jon Mease](jon.mease@gmail.com),
+           [Julia Signell](jsignell@gmail.com),
+- Status: Active
+- Type: Standards
+- Content-Type: text/markdown
+- Created: 2019-07-19
+
+## Abstract
+This spec describes a proposed shared API for accessing objects from various plotting libraries.
+
+In this notebook we will use **figure** to refer to any object that has a visible representation but which can also be saved, exported, etc. This includes generic visualizations or animation that may or may not be a "plot" (axes, ticks, etc)?
+
+When detailing a method, this spec uses FIGURE to refer to any specific figure object and LIBRARY to refer to any specific visualization library.
+
+## If a method doesn't make sense
+If a given operation doesn't make sense for that library, then it can satisfy the spec by simply having that method return a message to that effect ("JSON support not available in LIBRARY").
+
+## Jupyter methods
+Every library that supports rendering in Jupyter should support the various IPython rich display methods, i.e. `repr_html`, `repr_png`, `_ipython_display_`.
+
+## `.show()`
+Every figure should have the ability to render itself.
+
+### Proposal
+ 1) The top-level figure class has a `.show()` method that can be called without arguments to display the figure as a side-effect.
+ 2) The optional `renderer` kwarg can be used to override the current default renderer. e.g.:
+    - `FIGURE.show(renderer='png')` to display the figure as a static png image/
+    - `FIGURE.show(renderer='browser')` to display the figure in a browser tab. This works in non-jupyter/IPython contexts.
+
+## `.save()` method
+Every figure should have the knowledge of how to export itself to a file on disk.
+
+> #### What currently exists
+>  - matplotlib has `.savefig` which saves the current figure, with options `fname` and `format` among others.
+> - plotly has `write_*` methods which write the figure to a `file` (or writable object) and return `None`.
+> - bokeh has top level `export_*` methods for each output format as well as `save` (only outputs html) which both take a FIGURE as the arg.
+
+### Proposal
+Figures should have `FIGURE.save()` method for exporting.
+
+There are several kwargs that `save` should include:
+
+ 1) `file`: should be a file object or optionally a path to a file. If a file path is used rather than a file object and that is not supported, the library should raise a sensible error. Can default to some user configured value.
+ 2) `output`: should be a file format that the figure can be exported to i.e. `'png'`, `'json'`, `'html'`. The library should set a default output and/or allow the user to configure the output. For instance matplotlib does:
```

Comment:
```suggestion
 2) `output`: should be a file format that the figure can be exported to i.e. `'png'`, `'jpg'`, `'json'`, `'html'`. The library should set a default output and/or allow the user to configure the output. For instance matplotlib does:
```

(Added jpg explicitly here to avoid having some libraries choose 'jpg' and others 'jpeg' for the same operation.)